ndcube/ndcube.py

```diff
@@ -952,3 +962,234 @@ def to(self, new_unit, **kwargs):
+                 for i in range(n_pixels_per_bin):
+                     new_uncert += uncertainty.array[i]
+                 # Alternatively: new_uncerts = uncertainty.array.sum(axis=0)
+                 return type(uncertainty)(new_uncert)  # Convert to original uncert type and return.
+        """
+        # Sanitize input.
+        if new_unit is None:
+            new_unit = self.unit
+        # Make sure the input bin dimensions are integers.
+        bin_shape = np.rint(bin_shape).astype(int)
+        offsets = (bin_shape - 1) / 2
+        if all(bin_shape == 1):
+            return self
+        # Ensure bin_size has right number of entries and each entry is an
+        # integer fraction of the array shape in each dimension.
+        data_shape = self.dimensions.value.astype(int)
+        naxes = len(data_shape)
+        if len(bin_shape) != naxes:
+            raise ValueError("bin_shape must have an entry for each array axis.")
+        if (np.mod(data_shape, bin_shape) != 0).any():
+            raise ValueError(
+                "bin shape must be an integer fraction of the data shape in each dimension. "
+                f"data shape: {data_shape};  bin shape: {bin_shape}")
+
+        # Reshape array so odd dimensions represent pixels to be binned
+        # then apply function over those axes.
+        m = None if (self.mask is None or self.mask is False or use_masked_values) else self.mask
+        data = self.data
+        if m is None:
+            data = self.data
+        else:
+            masked_type = np.ma.masked_array
+            for array_type, m_type in ARRAY_MASK_MAP.items():
+                if isinstance(self.data, array_type):
+                    masked_type = m_type
+                    break
+            else:
+                warn.warning("data and mask arrays of different or unrecognized types. "
+                             "Casting them into a numpy masked array.")
+            data = masked_type(self.data, m)
+        reshape = np.empty(data_shape.size + bin_shape.size, dtype=int)
+        new_shape = (data_shape / bin_shape).astype(int)
+        reshape[0::2] = new_shape
+        reshape[1::2] = bin_shape
+        reshape = tuple(reshape)
+        reshaped_data = data.reshape(reshape)
+        operation_axes = tuple(range(len(reshape) - 1, 0, -2))
+        new_data = operation(reshaped_data, axis=operation_axes)
+        if isinstance(new_data, ARRAY_MASK_MAP[np.ndarray]):
+            new_data = new_data.data
+        if handle_mask is None:
+            new_mask = None
+        elif isinstance(self.mask, (type(None), bool)):  # Preserve original mask type.
+            new_mask = self.mask
+        else:
+            reshaped_mask = self.mask.reshape(reshape)
+            new_mask = handle_mask(reshaped_mask, axis=operation_axes)
+
+        # Propagate uncertainties is propagate_uncertainties kwarg set.
+        new_uncertainty = None
+        if propagate_uncertainties is False:
+            pass
```

Comment:
I am not a fan of this pattern.

Can we just put all this in a `if propagate_uncertainties` block?